traitsui/qt4/date_editor.py

```diff
@@ -154,14 +164,60 @@ def update_editor(self):
     def update_object(self, q_date):
         """ Handles the user entering input data in the edit control.
         """
-        year = q_date.year()
-        month = q_date.month()
-        day = q_date.day()
-        try:
-            self.value = datetime.date(year, month, day)
-        except ValueError:
-            print('Invalid date:', year, month, day)
-            raise
+        value = datetime.date(q_date.year(), q_date.month(), q_date.day())
+        if self.factory.multi_select:
+            if value in self.value:
+                self.unselect_date(value)
+            else:
+                self.select_date(value)
+
+            self.value = sorted(self._selected)
```

Comment:
Not sure if sorting the date is weird or not.  We could maintain the order of the dates to be in the same order as how the user selects them.  It is also somewhat inefficient that we keep sorting the list of date. @corranwebster What do you think?

Reply:
Sorting is the right behaviour for the date rage (there is a definite start and end, and a 'backwards' selection is likely to be more confusing for users even if it is theoretically more pure), and sorting is _probably_ the right behaviour for the multi-select (although it is a bit less obvious if the data coming in from the traits side is unsorted - I can see an argument that it maybe should be optional but on by default).

Timsort is fairly efficient for almost-sorted lists, if I recall, so the sorting overhead is likely not that much (particularly when the number of selected dates is likely to be small).

Reply:
Thanks!  I will keep the sort here then.